### ISSUE TITLE
Suffix Array consumed time

### ISSUE THREAD

**Issue by kimcando** [ChenghaoMou/text-dedup#22]
Hi, there

**Backgrounds:** I want to deduplicate the large volume data with suffix array algorithms.

**Progress:**
* My pure dataset size is about 465GB(temp_text.txt disk size)
* If following the code without any code edition, the output of the first step of "Suffix Array" is temp_text.txt.part.{string_num}-{next_chunk_string_num}, 
temp_text.txt.part.{string_num}-{next_chunk_string_num}.table.bin
* After the constructing binary table data, which is the first step of suffix array, merging suffix tree is now going on.

**Problem:**
* I found merging suffix tree is taking unexpected time. For instance, the first step took about less than 12 hours. But the merging step has been going on more than 3 days but if i calculate the expected time, it needs12 days more. What's more, there are more steps(e.g, merging individual tables, Doing selfsimilar, restore, deduplicate) to complete the whole process and I can't guess the expected time.

**Question:**
* How much time required to deduplicate 465GB data with the provided suffix array algorithm?
* Is there anything I can try to reduce the expected complete time?

thank you in advance for sharing the sources.
Cheers!

**Comment by ChenghaoMou** [comment]
Thanks for the question! Could you share your hardware details as well?

Regarding the time complexity, they are described in detail from the original repo here: https://github.com/google-research/deduplicate-text-datasets. But it is affected by the hardware as well. In particular, merging can be sped up by increasing num threads. 

According to their study:
```
If you want to deduplicate something the size of C4 (~300GB) you will want a machine with as many cores as you can get (we used 96 cores) and >600GB of RAM. You will also need >1TB hard drive space.
```
You would need something very similar to process your dataset in reasonable time. If I remember correctly, it should take less than a day in total for that scale of data, given proper compute.

**Comment by kimcando** [comment]
Actually, my hardware spec seems way enough to deal with the data.
* CPU core: 128
* RAM: 992 GB( while processing first step, it rans out of memory so I set 1TB swap memory as well)
* disk : 10TB

For the dataset details:
* temp_text.txt : 465G
* I have 100 `temp_text.txt.part.*` & and `temp_text.txt.part.*.table.bin` files respectively. Each *.table.bin file is about 24GB.

The following is my log..
```
[2023-06-19 14:29:37,273][INFO|make_suffix_array.py:85] >> Merging suffix trees
[2023-06-19 14:29:37,459][INFO|make_suffix_array.py:90] >> ./target/debug/dedup_dataset merge --output-file tmp/out.table.bin --suffix-path output/temp_text.txt.part.0-4988494083 --suffix-path output/temp_text.txt.part.4988394083-9976888166 --suffix-path output/temp_text.txt.part.9976788166-14965282249 --suffix-path output/temp_text.txt.part.14965182249-19953676332 --suffix-path output/temp_text.txt.part.19953576332-24942070415 --suffix-path output/temp_text.txt.part.24941970415-29930464498 --suffix-path output/temp_text.txt.part.29930364498-34918858581 --suffix-path output/temp_text.txt.part.34918758581-39907252664 --suffix-path output/temp_text.txt.part.39907152664-44895646747 --suffix-path output/temp_text.txt.part.44895546747-49884040830 --suffix-path output/temp_text.txt.part.49883940830-54872434913 --suffix-path output/temp_text.txt.part.54872334913-59860828996 --suffix-path output/temp_text.txt.part.59860728996-64849223079 --suffix-path output/temp_text.txt.part.64849123079-69837617162 --suffix-path output/temp_text.txt.part.69837517162-74826011245 --suffix-path output/temp_text.txt.part.74825911245-79814405328 --suffix-path output/temp_text.txt.part.79814305328-84802799411 --suffix-path output/temp_text.txt.part.84802699411-89791193494 --suffix-path output/temp_text.txt.part.89791093494-94779587577 --suffix-path output/temp_text.txt.part.94779487577-99767981660 --suffix-path output/temp_text.txt.part.99767881660-104756375743 --suffix-path output/temp_text.txt.part.104756275743-109744769826 --suffix-path output/temp_text.txt.part.109744669826-114733163909 --suffix-path output/temp_text.txt.part.114733063909-119721557992 --suffix-path output/temp_text.txt.part.119721457992-124709952075 --suffix-path output/temp_text.txt.part.124709852075-129698346158 --suffix-path output/temp_text.txt.part.129698246158-134686740241 --suffix-path output/temp_text.txt.part.134686640241-139675134324 --suffix-path output/temp_text.txt.part.139675034324-144663528407 --suffix-path output/temp_text.txt.part.144663428407-149651922490 --suffix-path output/temp_text.txt.part.149651822490-154640316573 --suffix-path output/temp_text.txt.part.154640216573-159628710656 --suffix-path output/temp_text.txt.part.159628610656-164617104739 --suffix-path output/temp_text.txt.part.164617004739-169605498822 --suffix-path output/temp_text.txt.part.169605398822-174593892905 --suffix-path output/temp_text.txt.part.174593792905-179582286988 --suffix-path output/temp_text.txt.part.179582186988-184570681071 --suffix-path output/temp_text.txt.part.184570581071-189559075154 --suffix-path output/temp_text.txt.part.189558975154-194547469237 --suffix-path output/temp_text.txt.part.194547369237-199535863320 --suffix-path output/temp_text.txt.part.199535763320-204524257403 --suffix-path output/temp_text.txt.part.204524157403-209512651486 --suffix-path output/temp_text.txt.part.209512551486-214501045569 --suffix-path output/temp_text.txt.part.214500945569-219489439652 --suffix-path output/temp_text.txt.part.219489339652-224477833735 --suffix-path output/temp_text.txt.part.224477733735-229466227818 --suffix-path output/temp_text.txt.part.229466127818-234454621901 --suffix-path output/temp_text.txt.part.234454521901-239443015984 --suffix-path output/temp_text.txt.part.239442915984-244431410067 --suffix-path output/temp_text.txt.part.244431310067-249419804150 --suffix-path output/temp_text.txt.part.249419704150-254408198233 --suffix-path output/temp_text.txt.part.254408098233-259396592316 --suffix-path output/temp_text.txt.part.259396492316-264384986399 --suffix-path output/temp_text.txt.part.264384886399-269373380482 --suffix-path output/temp_text.txt.part.269373280482-274361774565 --suffix-path output/temp_text.txt.part.274361674565-279350168648 --suffix-path output/temp_text.txt.part.279350068648-284338562731 --suffix-path out
put/temp_text.txt.part.284338462731-289326956814 --suffix-path output/temp_text.txt.part.289326856814-294315350897 --suffix-path output/temp_text.txt.part.294315250897-299303744980 --suffix-path output/temp_text.txt.part.299303644980-304292139063 --suffix-path output/temp_text.txt.part.304292039063-309280533146 --suffix-path output/temp_text.txt.part.309280433146-314268927229 --suffix-path output/temp_text.txt.part.314268827229-319257321312 --suffix-path output/temp_text.txt.part.319257221312-324245715395 --suffix-path output/temp_text.txt.part.324245615395-329234109478 --suffix-path output/temp_text.txt.part.329234009478-334222503561 --suffix-path output/temp_text.txt.part.334222403561-339210897644 --suffix-path output/temp_text.txt.part.339210797644-344199291727 --suffix-path output/temp_text.txt.part.344199191727-349187685810 --suffix-path output/temp_text.txt.part.349187585810-354176079893 --suffix-path output/temp_text.txt.part.354175979893-359164473976 --suffix-path output/temp_text.txt.part.359164373976-364152868059 --suffix-path output/temp_text.txt.part.364152768059-369141262142 --suffix-path output/temp_text.txt.part.369141162142-374129656225 --suffix-path output/temp_text.txt.part.374129556225-379118050308 --suffix-path output/temp_text.txt.part.379117950308-384106444391 --suffix-path output/temp_text.txt.part.384106344391-389094838474 --suffix-path output/temp_text.txt.part.389094738474-394083232557 --suffix-path output/temp_text.txt.part.394083132557-399071626640 --suffix-path output/temp_text.txt.part.399071526640-404060020723 --suffix-path output/temp_text.txt.part.404059920723-409048414806 --suffix-path output/temp_text.txt.part.409048314806-414036808889 --suffix-path output/temp_text.txt.part.414036708889-419025202972 --suffix-path output/temp_text.txt.part.419025102972-424013597055 --suffix-path output/temp_text.txt.part.424013497055-429001991138 --suffix-path output/temp_text.txt.part.429001891138-433990385221 --suffix-path output/temp_text.txt.part.433990285221-438978779304 --suffix-path output/temp_text.txt.part.438978679304-443967173387 --suffix-path output/temp_text.txt.part.443967073387-448955567470 --suffix-path output/temp_text.txt.part.448955467470-453943961553 --suffix-path output/temp_text.txt.part.453943861553-458932355636 --suffix-path output/temp_text.txt.part.458932255636-463920749719 --suffix-path output/temp_text.txt.part.463920649719-468909143802 --suffix-path output/temp_text.txt.part.468909043802-473897537885 --suffix-path output/temp_text.txt.part.473897437885-478885931968 --suffix-path output/temp_text.txt.part.478885831968-483874326051 --suffix-path output/temp_text.txt.part.483874226051-488862720134 --suffix-path output/temp_text.txt.part.488862620134-493851114217 --suffix-path output/temp_text.txt.part.493851014217-498839408394 --num-threads 128
```

**Comment by ChenghaoMou** [comment]
I will have to look into their rust code/repo. Meanwhile, could you run this on a much smaller dataset for a sanity check?

**Comment by ChenghaoMou** [comment]
Could you run your deduplication directly with their script and see how long it takes? https://github.com/google-research/deduplicate-text-datasets/blob/master/scripts/deduplicate_single_file.sh

My python script is only a wrapper on top of that script, so I think there might be some regression issues.

According to their paper:
> We run our algorithm on a single VM on the cloud with 96 cores and 768GB of memory. Our algorithm is efficient, for example processing the Wiki-40B training set (3 million examples containing 4GB of text) in 2.3 minutes wall-clock time (2.1 CPU-hours of work).
The 350GB C4 dataset takes under 12 hours (wall clock) to build a suffix array; although we are still memory constrained and so this corresponds to ∼ 1000 CPU-hours. Once the suffix array has been constructed, it takes under an hour to deduplicate the C4 dataset.

It should be able to handle 400GB of text in a day.

**Comment by ChenghaoMou** [comment]
I think this might be related: 

https://github.com/google-research/deduplicate-text-datasets/pull/22
https://github.com/google-research/deduplicate-text-datasets/issues/19

The bug might make the merge operation single-threaded. Applying that change and rebuild the binaries (`cargo build`) should help the issue.

**Comment by kimcando** [comment]
Thank you for the kind reply!
I will try the flow and share the result!!

**Comment by leoMesss** [comment]
Have you solved your problem now? i have tried the fix in [#19 ](https://github.com/google-research/deduplicate-text-datasets/issues/19), and it seems mutiple threads worked correctly.i have 128 CPU cores, and the merging suffix tree still cost much longer time compared to their paper, my dataset is about 15GB, and the merging part cost more than 3 hours(And it's still not finished)

**Comment by ChenghaoMou** [comment]
> Have you solved your problem now? i have tried the fix in [#19 ](https://github.com/google-research/deduplicate-text-datasets/issues/19), and it seems mutiple threads worked correctly.i have 128 CPU cores, and the merging suffix tree still cost much longer time compared to their paper, my dataset is about 15GB, and the merging part cost more than 3 hours(And it's still not finished)

It is unlikely that it will take that long. Could you remove all the cache files and try one more time? Feel free to create an issue if it persists.

**Comment by leoMesss** [comment]
I think there are some long duplicates in my dataset (more than 50000000 bytes as the warning said), this might be the reason why it took so much time.  The dataset is a Chinese Dataset called Wudao200G，they claimed they had already cleaned it, so i didn't realize the problem at first time, after i tried some other datasets(the code worked pretty well), i think the problem is the dataset.

**Comment by leoMesss** [comment]
> > Have you solved your problem now? i have tried the fix in [#19 ](https://github.com/google-research/deduplicate-text-datasets/issues/19), and it seems mutiple threads worked correctly.i have 128 CPU cores, and the merging suffix tree still cost much longer time compared to their paper, my dataset is about 15GB, and the merging part cost more than 3 hours(And it's still not finished)
> 
> It is unlikely that it will take that long. Could you remove all the cache files and try one more time? Feel free to create an issue if it persists.

I think there are some long duplicates in my dataset (more than 50000000 bytes as the warning said), this might be the reason why it took so much time. The dataset is a Chinese Dataset called Wudao200G，they claimed they had already cleaned it, so i didn't realize the problem at first time, after i tried some other datasets(the code worked pretty well), i think the problem is the dataset.
